src/main/java/io/stargate/sgv2/jsonapi/service/operation/model/impl/DeleteOperation.java

```diff
@@ -27,58 +29,73 @@ public record DeleteOperation(
-                  // Below conditionality is because we read up to deleteLimit +1 record.
-                  if (totalCount.get() + docs.size() <= deleteLimit) {
-                    totalCount.addAndGet(docs.size());
-                    return Multi.createFrom().items(docs.stream());
-                  } else {
-                    int needed = deleteLimit - totalCount.get();
-                    totalCount.addAndGet(needed);
-                    moreData.set(true);
-                    return Multi.createFrom()
-                        .items(findResponse.docs().subList(0, needed).stream());
-                  }
-                })
-            .concatenate()
-            .onItem()
-            .transformToUniAndConcatenate(
-                readDocument -> deleteDocument(queryExecutor, delete, readDocument))
-            .collect()
-            .in(
-                AtomicInteger::new,
-                (atomicCounter, flag) -> {
-                  if (flag) {
-                    atomicCounter.incrementAndGet();
-                  }
-                });
+    // Read the required records to be deleted
+    return Multi.createBy()
+        .repeating()
+        .uni(
+            () -> new AtomicReference<String>(null),
+            stateRef -> {
+              Uni<ReadOperation.FindResponse> docsToDelete =
+                  readOperation().getDocuments(queryExecutor, stateRef.get(), null);
+              return docsToDelete
+                  .onItem()
+                  .invoke(findResponse -> stateRef.set(findResponse.pagingState()));
+            })
 
-    return counter
+        // Documents read until pagingState available, max records read is deleteLimit + 1
+        .whilst(findResponse -> findResponse.pagingState() != null)
+
+        // Get the deleteLimit # of documents to be delete and set moreData flag true if extra
+        // document is read.
+        .onItem()
+        .transformToMulti(
+            findResponse -> {
+              final List<ReadDocument> docs = findResponse.docs();
+              // Below conditionality is because we read up to deleteLimit +1 record.
+              if (totalCount.get() + docs.size() <= deleteLimit) {
+                totalCount.addAndGet(docs.size());
+                return Multi.createFrom().items(docs.stream());
+              } else {
+                int needed = deleteLimit - totalCount.get();
+                totalCount.addAndGet(needed);
+                moreData.set(true);
+                return Multi.createFrom().items(findResponse.docs().subList(0, needed).stream());
+              }
+            })
+        .concatenate()
+
+        // Run delete for selected documents and retry in case of
+        .onItem()
+        .transformToUniAndConcatenate(
+            document -> {
+              AtomicInteger retryAttempt = new AtomicInteger(0);
+              return deleteDocument(queryExecutor, delete, document, retryAttempt)
```

Comment:
this is not good imo.. The delete document should not do the re-read, it should be seomthing like:

```
deleteDocument(..)

   .onFailure(jwtPredicate)
   .switchTo(() -> {
        readDocumentAgain(queryExecutor, document)

           .flatMap(document -> deleteDocument(..))

           .onFailure(jwtPredicate)
           .retry().max(maxLwtRetries - 1)
    })
```

You need to avoid using Atomic stuff, and side effects in general.. Plus the suggested code is way simpler and smaller..

Reply:
Sorry, mixed the frameworks.. Here it's

```
.onFailure(..).recoverWithUni(..)
```